### DISCUSSION TITLE
Added: LDM Notes for September 1st, 2021

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#5141] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-09-01.md

## Agenda

1. Lambda expression conversions to `Delegate`
2. C# 11 Initialization Triage
    1. Required properties
    2. Primary constructors
    3. Immutable collection initializers

Again, same warning to readers to not assume that because we talked about a feature in these triage meetings, it will or will not be in C# 11.

**Comment by MgSam** [comment]
> Do we actually want to do this feature because customers are asking for it, or do we just think it's interesting as an LDM?

Funny to see this quote in the same meeting in which primary constructors are revived yet again. Has anyone ever actually asked for this feature?

Separately, I think the type system features (like shapes, statics on interfaces, etc) are a lot more bang-for-the-buck than adding yet more syntax sugars to do the same things we can already do, which is what a lot of the proposed features are. 

The shapes proposal is nearly 5 years old already. Would be nice to see some real movement on it rather than just oblique references that you ‘could’ look at doing something there someday. 

Put another way, it’s almost 2022 and we still can’t make extension methods to static classes (Math, etc.)

**Reply by jnm2** [reply]
I want primary constructors very much.

**Reply by CyrusNajmabadi** [reply]
> Has anyone ever actually asked for this feature?

Yes.  *Many* (including LDM members, as well as many first, second and third parties).

**Reply by 333fred** [reply]
> Funny to see this quote in the same meeting in which primary constructors are revived yet again. Has anyone ever actually asked for this feature?

That quote was not related to primary constructors. It was related to record/class cross inheritance.

**Reply by MgSam** [reply]
> > Funny to see this quote in the same meeting in which primary constructors are revived yet again. Has anyone ever actually asked for this feature?
> 
> That quote was not related to primary constructors. It was related to record/class cross inheritance.

I figured; thus my comment on the irony. Mads’ last proposal had pretty tepid support in the discussions and the original proposal for C# 6 died from overwhelming opposition by the community. Just seems odd to me that given all that this still seems like a feature worth spending limited design time on.

**Reply by CyrusNajmabadi** [reply]
> Just seems odd to me that given all that this still seems like a feature worth spending limited design time on.

We can only make that determination through discussion.

**Reply by HaloFour** [reply]
Mads last proposal had tepid support because it was a departure from the syntax established in records and would make the simpler cases much more convoluted.  The original proposal on CodePlex only gained opposition because it attempted to achieve parity with normal constructors and in doing so had become a mess.  Primary constructors, as already implemented in records and as implemented in so many other languages would be a very welcome addition to C#.
